std/selector/multiplexer.go

```diff
@@ -32,76 +30,112 @@ func GetHints() []solver.Hint {
 	return []solver.Hint{stepOutput, muxIndicators, mapIndicators}
 }
 
-// Map is a key value associative array: the output will be values[i] such that keys[i] == queryKey. If keys does not
-// contain queryKey, no proofs can be generated. If keys has more than one key that equals to queryKey, the output will
-// be undefined, and the output could be a linear combination of all the corresponding values with that queryKey.
+// Map is a key value associative array: the output will be values[i] such that
+// keys[i] == queryKey. If keys does not contain queryKey, no proofs can be
+// generated. If keys has more than one key that equals to queryKey, the output
+// will be undefined, and the output could be any linear combination of all the
+// corresponding values with that queryKey.
 //
-// In case keys and values do not have the same length, this function will panic.
+// In case keys and values do not have the same length, this function will
+// panic.
 func Map(api frontend.API, queryKey frontend.Variable,
 	keys []frontend.Variable, values []frontend.Variable) frontend.Variable {
-	// we don't need this check, but we added it to produce more informative errors and disallow
-	// len(keys) < len(values) which is supported by generateSelector.
+	// we don't need this check, but we added it to produce more informative errors
+	// and disallow len(keys) < len(values) which is supported by generateSelector.
 	if len(keys) != len(values) {
 		panic("The number of keys and values must be equal")
 	}
-	return generateSelector(api, false, queryKey, keys, values)
+	return dotProduct(api, values, KeyDecoder(api, queryKey, keys))
 }
 
-// Mux is an n to 1 multiplexer: out = inputs[sel]. In other words, it selects exactly one of its
-// inputs based on sel. The index of inputs starts from zero.
+// Mux is an n to 1 multiplexer: out = inputs[sel]. In other words, it selects
+// exactly one of its inputs based on sel. The index of inputs starts from zero.
 //
-// sel needs to be between 0 and n - 1 (inclusive), where n is the number of inputs, otherwise the proof will fail.
+// sel needs to be between 0 and n - 1 (inclusive), where n is the number of
+// inputs, otherwise the proof will fail.
 func Mux(api frontend.API, sel frontend.Variable, inputs ...frontend.Variable) frontend.Variable {
-	return generateSelector(api, true, sel, nil, inputs)
+	// we use BinaryMux when len(inputs) is a power of 2 and its math.Log2 is an
+	// integer.
+	if log := math.Log2(float64(len(inputs))); log == math.Trunc(log) {
```

Comment:
Could also do `bits.Onescount(len(inputs)) == 1`, see https://pkg.go.dev/math/bits#OnesCount. I would always prefer to use floats due to imprecision.

Reply:
Yes, that's a good idea.